samples/subsys/ipc/openamp/src/main.c

```diff
@@ -0,0 +1,294 @@
+		printk("Couldn't register shared memory device: %d\n", status);
+		return;
+	}
+
+	status = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (status != 0) {
+		printk("metal_device_open failed: %d\n", status);
+		return;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (io == NULL) {
+		printk("metal_device_io_region failed to get region\n");
+		return;
+	}
+
+	/* setup IPM */
+	ipm_handle = device_get_binding("MAILBOX_0");
+	if (ipm_handle == NULL) {
+		printk("device_get_binding failed to find device\n");
+		return;
+	}
+
+	ipm_register_callback(ipm_handle, platform_ipm_callback, NULL);
+
+	ipm_set_enabled(ipm_handle, 1);
+	if (status != 0) {
+		printk("ipm_set_enabled failed\n");
+		return;
+	}
+
+	/* setup vdev */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (vq[0] == NULL) {
+		printk("virtqueue_allocate failed to alloc vq[0]\n");
+		return;
+	}
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (vq[1] == NULL) {
+		printk("virtqueue_allocate failed to alloc vq[1]\n");
+		return;
+	}
+
+	vdev.role = RPMSG_MASTER;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.vrings_info = &rvrings[0];
+
+	/* setup rvdev */
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	status = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+	if (status != 0) {
+		printk("rpmsg_init_vdev failed %d\n", status);
+		return;
+	}
+
+	/* Since we are using name service, we need to wait for a response
+	 * from NS setup and than we need to process it
+	 */
+	k_sem_take(&data_sem, K_FOREVER);
+	virtqueue_notification(vq[0]);
+
+	/* Wait til nameservice ep is setup */
+	k_sem_take(&ept_sem, K_FOREVER);
+
+	while (message < 100) {
+		status = send_message(message);
+		if (status < 0) {
+			printk("send_message(%d) failed with status %d\n",
+			       message, status);
+			goto _cleanup;
+		}
+
+		message = receive_message();
+		printk("Master core received a message: %d\n", message);
+
+		message++;
+	}
+
+_cleanup:
+	rpmsg_deinit_vdev(&rvdev);
+	metal_finish();
+
+	printk("OpenAMP demo ended.\n");
+}
+
+void main(void)
+{
+	printk("Starting application thread!\n");
+	k_thread_create(&thread_data, thread_stack, APP_TASK_STACK_SIZE,
+			(k_thread_entry_t)app_task,
+			NULL, NULL, NULL, K_PRIO_COOP(7), 0, 0);
+}
+
+/* Make sure we clear out the status flag very early (before we bringup the
+ * secondary core) so the secondary core see's the proper status
+ */
+int init_status_flag(struct device *arg)
+{
+	virtio_set_status(NULL, 0);
+
+	return 0;
+}
+
+SYS_INIT(init_status_flag, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```

Comment:
Only see init_status_flag() called from master, it suggests master needs to load first, better to have some comments on the init_status_flags.